client/src/components/Project/Project.js

```diff
@@ -0,0 +1,421 @@
+import Grid from '@material-ui/core/Grid'
+import checkboxAdnRadioStyle from '../../assets/jss/material-dashboard-react/checkboxAdnRadioStyle.js'
+import { Formik, Form, ErrorMessage } from 'formik'
+import { useSelector, useDispatch } from 'react-redux'
+import {
+  addNewProject,
+  clearProjectMsg,
+  updateProject
+} from '../../actions/projectAction'
+import * as Yup from 'yup'
+
+import {
+  MuiPickersUtilsProvider,
+  KeyboardDatePicker
+} from '@material-ui/pickers'
+
+const styles = {
+  ...checkboxAdnRadioStyle,
+  cardCategoryWhite: {
+    color: 'rgba(255,255,255,.62)',
+    margin: '0',
+    fontSize: '14px',
+    marginTop: '0',
+    marginBottom: '0'
+  },
+  cardTitleWhite: {
+    color: '#FFFFFF',
+    marginTop: '0px',
+    minHeight: 'auto',
+    fontWeight: '300',
+    fontFamily: "'Roboto', 'Helvetica', 'Arial', sans-serif",
+    marginBottom: '3px',
+    textDecoration: 'none'
+  },
+  root: {
+    display: 'flex',
+    flexWrap: 'wrap'
+  },
+  formControl: {
+    margin: 11,
+    minWidth: 120,
+    wrap: 'nowrap',
+    fullWidth: 'true',
+    display: 'flex'
+  },
+  selectEmpty: {
+    marginTop: 10
+  },
+  error: {
+    color: 'red'
+  },
+  dateStyle: {
+    paddingLeft: 11,
+    paddingRight: 11
+  }
+}
+
+const useStyles = makeStyles(styles)
+const Project = props => {
+  const { setPageView, projectToUpdate } = props
+  const classes = useStyles()
+  const { addToast } = useToasts()
+  const dispatch = useDispatch()
+
+  const error = useSelector(state => state.projectReducer.error)
+  const addNewProjectStatus = useSelector(
+    state => state.projectReducer.addNewProjectStatus
+  )
+  const updateProjectStatus = useSelector(
+    state => state.projectReducer.updateProjectStatus
+  )
+  const updateProjectError = useSelector(
+    state => state.projectReducer.updateProjectError
+  )
+  const projectForm = useRef(null)
+
+  useEffect(() => {
+    if (addNewProjectStatus) {
+      addToast(addNewProjectStatus, {
+        appearance: 'success',
+        autoDismiss: true
+      })
+      projectForm.current.reset()
+      dispatch(clearProjectMsg())
+      setPageView('projectListing')
+    }
+    if (updateProjectStatus) {
+      addToast(updateProjectStatus, {
+        appearance: 'success',
+        autoDismiss: true
+      })
+      projectForm.current.reset()
+      dispatch(clearProjectMsg())
+      if (props) props.setUpdateAction()
+    }
+  }, [addNewProjectStatus, updateProjectStatus, addToast])
+
+  useEffect(() => {
+    if (error) {
+      addToast(error, { appearance: 'error', autoDismiss: true })
+      dispatch(clearProjectMsg())
+    }
+
+    if (updateProjectError) {
+      addToast(updateProjectError, { appearance: 'error', autoDismiss: true })
+      dispatch(clearProjectMsg())
+    }
+  }, [error, updateProjectError, addToast])
+
+  const submitFormValues = values => {
+    if (projectToUpdate) {
+      const id = projectToUpdate[0]._id
+      dispatch(updateProject(values, id))
```

Comment:
can you optimize something like this? `dispatch(expression?updateProject(values, id):addNewProject(values))`

Reply:
Done